xbmc/threads/Thread.cpp

```diff
@@ -102,6 +102,7 @@ void CThread::Create(bool bAutoDelete, unsigned stacksize)
 
 bool CThread::IsRunning() const
 {
+  CSingleLock lock(m_CriticalSection);
```

Comment:
As explained FernetMenta all those things you are saying have nothing to do with why the lock/atomic is needed here. To re-cite: "When an evaluation of an expression writes to a memory location and another evaluation reads [the program has a data race]. If a data race occurs, the behavior of the program is undefined". Ie. both your examples have undefined behaviour at the 'if' where IsRunning is called. it does not matter what comes after.